system/trace/trace.c

```diff
@@ -0,0 +1,812 @@
+
+static void notectl_enable(int flag)
+{
+  struct note_filter_mode_s mode;
+
+#ifdef CONFIG_BUILD_FLAT
+  sched_note_filter_mode(&mode, NULL);
+#else
+  ioctl(g_notectlfd, NOTECTL_GETMODE, (unsigned long)&mode);
+#endif
+
+  if (flag)
+    {
+      mode.flag |= NOTE_FILTER_MODE_FLAG_ENABLE;
+    }
+  else
+    {
+      mode.flag &= ~NOTE_FILTER_MODE_FLAG_ENABLE;
+    }
+
+#ifdef CONFIG_BUILD_FLAT
+  sched_note_filter_mode(NULL, &mode);
+#else
+  ioctl(g_notectlfd, NOTECTL_SETMODE, (unsigned long)&mode);
+#endif
+}
+
+/****************************************************************************
+ * Name: note_ioctl
+ ****************************************************************************/
+
+#ifdef CONFIG_DRIVER_NOTERAM
+static void note_ioctl(int cmd, unsigned long arg)
+{
+  int fd;
+
+  fd = open("/dev/note", O_RDONLY);
+  if (fd < 0)
+    {
+      fprintf(stderr,
+             "trace: cannot open /dev/note\n");
+      return;
+    }
+
+  ioctl(fd, cmd, arg);
+  close(fd);
+}
+#endif
+
+/****************************************************************************
+ * Name: trace_cmd_start
+ ****************************************************************************/
+
+static int trace_cmd_start(int index, int argc, char **argv)
+{
+  char *endptr;
+  int duration = 0;
+
+  /* Usage: trace start [<duration>] */
+
+  if (index < argc - 1)
+    {
+      index++;
+      duration = strtoul(argv[index], &endptr, 0);
+      if (!duration || endptr == argv[index] || *endptr != '\0')
+        {
+          fprintf(stderr,
+                  "trace start: invalid argument '%s'\n", argv[index]);
+          return ERROR;
+        }
+    }
+
+  /* Clear the trace buffer */
+
+#ifdef CONFIG_DRIVER_NOTERAM
+  note_ioctl(NOTERAM_CLEAR, 0);
+#endif
+
+  /* Start tracing */
+
+  notectl_enable(true);
+
+  if (duration > 0)
+    {
+      /* If <duration> is given, stop tracing after specified seconds. */
+
+      sleep(duration);
+      notectl_enable(false);
+    }
+
+  return index;
+}
+
+/****************************************************************************
+ * Name: trace_cmd_dump
+ ****************************************************************************/
+
+#ifdef CONFIG_DRIVER_NOTERAM
+static int trace_cmd_dump(int index, int argc, char **argv)
+{
+  FILE *out = stdout;
+  int ret;
+
+  /* Usage: trace dump [<filename>] */
+
+  /* If <filename> is '-' or not given, trace dump is displayed
+   * to stdout.
+   */
+
+  if (index < argc - 1)
+    {
+      index++;
+      if (strcmp(argv[index], "-") != 0)
+        {
+          /* If <filename> is given, open the file stream for output. */
+
+          out = fopen(argv[index], "w");
+          if (out == NULL)
+            {
+              fprintf(stderr,
+                      "trace dump: cannot open '%s'\n", argv[index]);
+              return ERROR;
+            }
+        }
+    }
+
+  /* Stop the tracing before dump */
+
+  notectl_enable(false);
```

Comment:
trace_dump() continues dumping until all data in the buffer is dumped. 
Because trace_dump() itself issues syscalls, if the trace is not stopped before dump, these are added in the buffer while dumping. So, once the dump is started, it never stops.